feat (#746): Реализована возможность чтения из скрипта данных, переданных через stdin

arkuznetsov opened the pull request:
Реализована возможность чтения из скрипта данных, переданных через stdin.

test.os
```bsl
Чтение = Новый ЧтениеТекста();
Чтение.Открыть(СтандартныйПотокВвода);
Сообщить(СокрЛП(Чтение.Прочитать()));
```

Запускаем:
```bat
> echo 1234 | oscript test.os
```

Вывод:
```
1234
```

Commit:
feat (#746): Реализована возможность чтения из скрипта данных, переданных через stdin.

Commit:
test: Поправлен тест.

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/SystemGlobalContext.cs:
```diff
@@ -38,7 +38,17 @@ public SystemGlobalContext()
+                ConsoleInput.CopyTo(StandartInput);
+                ConsoleInput.Close();
+            }
+            RegisterProperty("СтандартныйПотокВвода", () => StandartInput);
```
Давай сделаем это не глобальным свойством, а в классе Консоль? Не хочется занимать глобальный неймспейс без нужды... @nixel2007 @artbear @dmpas @Infactum - реквестирую консилиум.

arkuznetsov replied:
А в таком случае, как должно выглядеть обращение из скрипта?

nixel2007 replied:
я могу поделиться примерами из java.

Обычно работа с потоком ввода идет через объект `System.in`, который держит ссылку на `InputStream`, у которого есть read-методы по чтению сырых байтов. При этом есть класс `Console`, ссылка на который лежит там же в `System.console()`. Там методы уже покрасивее, есть readLine, format, readPassword и прочая красота.

При этом запись в stdout может выполняться как из `System.out` так и из `System.console()`.

System.in появился в 1.1, то есть грубо говоря на заре появления джавы. Console - уже в 1.6, то есть некоторого после взятия за голову.

Но на практике напрямую из потоков и/или консоли все равно никто не читает, а оборачивают обращения к System.in через вызовы `Scanner` (который умеет `readInt`, `readFloat` и прочее) либо в `InputStreamReader`.

---

Занимать глобальный нэймспэйс точно не хочется. Перенести поток в `Консоль` выглядит вполне логично и похоже на ту же джаву. Поток же все равно нельзя будет получить, если скрипт запущен не из терминала, да? помню, что обращение к `Консоль` падает при запуске в определенных вне-терминальных окружениях. Не будет ли здесь проблем?

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/SystemGlobalContext.cs:
```diff
@@ -38,7 +38,17 @@ public SystemGlobalContext()
+            MemoryStreamContext StandartInput = MemoryStreamContext.Constructor();
+            if (Console.IsInputRedirected)
+            {
+                GenericStream ConsoleInput = new GenericStream(Console.OpenStandardInput(), true);
```
Мне больше нравится API Console.OpenStandardInput() который вызывается, как метод-действие по требованию.

arkuznetsov replied:
Ты предлагаешь, сделать типа:
```bsl
Консоль = Новый Консоль();
Консоль.ПолучитьСтандартныйВвод();
```
?

EvilBeaver replied:
Да. Консоль.ОткрытьПотокВвода(), например

arkuznetsov commented:
Хм, чот на Linux не отрабатывает. :-(
@EvilBeaver @dmpas Что может быть?

EvilBeaver commented:
> Хм, чот на Linux не отрабатывает. :-(
> @EvilBeaver @dmpas Что может быть?

Все что угодно. Как именно оно не отрабатывает? Ругается на что-то? Есть тексты ошибок?

arkuznetsov commented:
> > Хм, чот на Linux не отрабатывает. :-(
> > @EvilBeaver @dmpas Что может быть?
> 
> Все что угодно. Как именно оно не отрабатывает? Ругается на что-то? Есть тексты ошибок?

Просто не получает данные с stdin. Без ошибок. Руками не запускал, просто тесты посмотрел.

EvilBeaver commented:
Я не знаю. Предлагаю копать дальше. Погуглить известные баги в mono или core (смотря под чем гоняешь тесты)

arkuznetsov commented:
> Я не знаю. Предлагаю копать дальше. Погуглить известные баги в mono или core (смотря под чем гоняешь тесты)

Да, конечно, буду ковырять. А тесты я на твоем Женькинсе смотрел :-).

Commit:
refactor: Получение stdin перенесено в объект ConsoleContext.
feat: ConsoleContext сделан синглтоном;
Добавлено глобальное свойство "Консоль".

arkuznetsov commented:
@EvilBeaver Переделал.

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/SystemGlobalContext.cs:
```diff
@@ -38,7 +38,7 @@ public SystemGlobalContext()
             RegisterProperty("FileStreams", () => FileStreams);
 
-			RegisterProperty("Символы", () => (IValue)Chars);
+            RegisterProperty("Символы", () => (IValue)Chars);
```
Вот сюда бы добавить свойство Консоль, чтобы был глобальный синглтон. @dmpas, что думаешь?

arkuznetsov replied:
> Вот сюда бы добавить свойство Консоль, чтобы был глобальный синглтон. @dmpas, что думаешь?

Уже сразу в тестах нарвался на 'Консоль = Новый Консоль()' боюсь, что много где может вылезти.

nixel2007 replied:
может не глобальный, а через враппер? типа Система?
я помню, что вы наоборот выносили из системной информации всякие методы и свойства наружу, но может просто было нехорошее имя?

EvilBeaver commented on src/ScriptEngine.HostedScript/Library/ConsoleContext.cs:
```diff
@@ -181,9 +208,18 @@ public void Beep()
-            return new ConsoleContext();
+            if (_console == null)
+            {
+                lock (_lock)
```
lock не нужен, см. статью Скита про синглоны. Нужно создать пустой статик конструктор и обычное статическое поле. https://csharpindepth.com/articles/singleton

Варианты 5 и 6

arkuznetsov replied:
Опять же оставил конструктор, чтобы не ломать совместимость. С lock посмотрю.

nixel2007 commented on src/ScriptEngine.HostedScript/Library/ConsoleContext.cs:
```diff
@@ -181,9 +208,18 @@ public void Beep()
         public static ConsoleContext Constructor()
         {
-            return new ConsoleContext();
+            if (_console == null)
```
если это попытка в DLC, то в мультитреде оно работает не так, как ты ожидаешь.

https://csharpindepth.com/articles/singleton

Commit:
Заготовка по #746

# Conflicts:
#	src/ScriptEngine.HostedScript/Library/ConsoleContext.cs

EvilBeaver committed:
fixed #583 fixed #746

EvilBeaver committed:
Ошибочно вкоммитил флаг ошибок

EvilBeaver committed:
Удален старый тест

arkuznetsov commented on src/ScriptEngine.HostedScript/Library/ConsoleContext.cs:
```diff
@@ -196,6 +171,24 @@ public IValue InputEncoding
+        /// Возвращает или задает кодировку консоли, используемую при чтении входных данных.
+        /// </summary>
+        /// <returns>КодировкаТекста</returns>
+        [ContextProperty("КодировкаВыходногоПотока", "InputEncoding")]
```
Наверное "КодировкаВходногоПотока".

EvilBeaver replied:
Нет, тут именно Выходного. Их там две. Но английский синоним забыл поменять после копипасты. Спасибо

arkuznetsov replied:
Угу, не туда посмотрел.

arkuznetsov committed:
Merge branch 'EvilBeaver:develop' into feature/get-stdin

Commit:
Проверка получения стандартного потока ввода на Linux

Commit:
Проверка получения стандартного потока ввода на Linux (2)

Commit:
Откат изменений для отладки перенаправления ввода.

Commit:
Исправлен тест на Linux
